[shuffler] IR

danking opened the pull request:
This implements the Shuffler IR, which has become these four nodes:
- ShuffleWith: owns the shuffle, passes the shuffle id to children as a binding.
- ShuffleWrite: writes a stream into the shuffler.
- ShufflePartitionBounds: gets the boundaries of approximately balanced partitions.
- ShuffleRead: reads a stream from the shuffler.

It adds `TShuffle` which represents a shuffle and contains all the EType, PType, and Type information necessary to implement all nodes (thanks @catoverdrive !). `TShuffle` is realized at runtime by a `PCanonicalShuffle(true)`, which is a straight copy of `PCanonicalBinary`. Unfortunately, it appears not possible to have two distinct virtual types whose run-time realization is the same physical type.

`testShuffleIR` is most of the work towards a lowering of TableKeyBy/TableSortBy, I believe.

Next Up:
- Actually implement "successful partition ids," which is ignored by the IR. The server doesn't even have functionality for this right now. Note there are FIXME comments about this in this PR. I will remove them in the next PR.

danking commented:
This whole stack of PRs is currently plagued by testng hell. I will resolve Monday. I assure the tests work when not run using the test jar. I'll probably back out my use of the buggy TestNG feature (it's in LSMSuite in a previous PR).

danking commented:
bump

danking commented:
bump

danking commented:
@cseed ready for review

cseed commented on hail/src/main/scala/is/hail/shuffler/ShuffleClient.scala:
```diff
@@ -15,46 +16,101 @@ object ShuffleClient {
+
+  def openConnection(cb: EmitCodeBuilder, typ: TShuffle): (SocketCode, InputBufferValue, OutputBufferValue, LoggerCode) = {
+    val socketLocal = cb.emb.newLocal[Socket]("shuffleClientSocket")
+    cb.append(socketLocal := codeSocket())
```
I think you can just write:

```
val socket = cb.newLocal("shuffleClientSocket", codeSocket())
```

This just declares a local with an initializer (at the end of the code builder).

This pattern shows up in a few places.

cseed commented on hail/src/main/scala/is/hail/shuffler/LoggerCode.scala:
```diff
@@ -0,0 +1,25 @@
+}
+
+
+class LoggerCode(
```
Small stylistic thing: we normally put Code first: CodeArray, CodeInt, RichCodeRegion, etc.

danking replied:
Actually, I went with `...Value` since they all require a value.

It seems somewhat bad that we've bifurcated into `P...Code` & `P...Value` and `Code...` as the two naming schemes. I don't much care, but it seems bad to perpetuate, by adding `Value...`? Or maybe better to keep prefix style for non-Hail-objects?

cseed commented on hail/src/main/scala/is/hail/shuffler/LoggerCode.scala:
```diff
@@ -0,0 +1,25 @@
+
+class LoggerCode(
+  code: Value[Logger]
+) extends Value[Logger] {
```
The other pattern we normally use is an implicit conversion so you can write `log: Value[Logger]` or `log: Code[Logger]` in the code, and still say `log.info()`.  See RichCodeRegion for example.

cseed commented on hail/src/main/scala/is/hail/shuffler/RegionCode.scala:
```diff
@@ -0,0 +1,11 @@
+import is.hail.types.physical._
+
+object RegionCode {
+  def pretty(pType: PType, off: Code[Long]): Code[String] =
```
There is already a `Region.pretty`.

cseed commented on hail/src/main/scala/is/hail/expr/ir/EmitStream.scala:
```diff
@@ -1429,6 +1437,148 @@ object EmitStream {
+                out.writeInt(nPartitions),
+                out.flush())
+              val close = Code(
+                log.info("CLNT get sending EOS"),
```
Thought we were dropping CLNT since the logger will have context.

cseed commented on hail/src/main/scala/is/hail/shuffler/ShuffleClient.scala:
```diff
@@ -15,46 +16,101 @@ object ShuffleClient {
+  }
+
+  def socket(): Socket = {
+    var host = System.getenv("SHUFFLER_SSL_CLIENT_HOST")
```
I assume you're going to switch this to use deploy config at some point?  If we need another deploy config setup for deploying locally we can add that.

danking replied:
Once we sort out TCP forwarding, yes, I will modify deploy config and modify this to use it.

cseed commented on hail/src/main/scala/is/hail/expr/ir/Requiredness.scala:
```diff
@@ -638,6 +647,20 @@ class Requiredness(val usesAndDefs: UsesAndDefs, ctx: ExecuteContext) {
       case BlockMatrixCollect(child) =>  // BlockMatrix elements are all required
+      case ShuffleWith(keyFields, rowType, rowEType, keyEType, name, writer, readers) =>
+        requiredness.union(lookup(writer).required)
+        requiredness.unionFrom(lookup(readers))
```
I'm wondering, do we want this level of generality of missingness for writer, id, etc.?  It seems like something is wrong if you ask for the partition bounds with a missing id, for example.  This is much more flexible than the existing code.

cseed replied:
What happens if one successful partition is missing?

cseed commented on hail/src/main/scala/is/hail/expr/ir/Requiredness.scala:
```diff
@@ -96,6 +96,14 @@ class Requiredness(val usesAndDefs: UsesAndDefs, ctx: ExecuteContext) {
+        uses.foreach { u => defs.bind(u, Array(req)) }
+        dependents.getOrElseUpdate(d, mutable.Set[RefEquality[BaseIR]]()) ++= uses
+      }
+    }
```
I can't figure out how this works.  I know what needs to happen: the ShuffeWrite needs need to determine the requiredness of the TShuffle, which in turn determines the requiredness of the read and partition bounds nodes.  I'm still a little confused by the implementation of Requiredness but I'll keep reading it over.

danking replied:
@catoverdrive you should probably review the changes to this file.

cseed replied:
I'll keep trying to understand this, but I would also appreciate @catoverdrive's review.

cseed commented on hail/src/main/scala/is/hail/shuffler/SameLengthByteArrayComparator.scala:
```diff
@@ -4,7 +4,8 @@ import java.util.Comparator
 class SameLengthByteArrayComparator extends Comparator[Array[Byte]] {
   def compare(l: Array[Byte], r: Array[Byte]): Int = {
-    assert(l.length == r.length)
+    if (!(l.length == r.length))
```
`!=`?

danking replied:
Sure, changed. I've been doing a lot of mechanical messageless assertion to if-throws and figured the `if(!(...))` pattern elucidated what's happening.

cseed commented on hail/src/main/scala/is/hail/shuffler/ShuffleClient.scala:
```diff
@@ -15,46 +16,102 @@ object ShuffleClient {
   private[this] val log = Logger.getLogger(getClass.getName())
 
   lazy val sslContext = {
```
My tls stuff is in now.  Do you want to use that?  `tls.getSSLContext`.

danking commented:
I attempted to use the TLS stuff and I've decided against it for this PR. It appears that the batch tests do not work locally at all. The whole essence of this PR is getting the shuffler IR tested and into the mainline. I started making the fixes necessary to support local testing of a local server and when that proved complicated investigated how BatchClientSuite works locally.

It seems that using the existing TLS stuff would require fixing all the TLS stuff to allow for at least local->remote testing, if not local->local testing.

This PR is already very complex, I'd like to get it merged so we can move forward separately with deploying and eventually harmonizing with the existing TLS infrastructure.

danking commented on hail/src/main/scala/is/hail/expr/ir/InferPType.scala:
```diff
@@ -265,13 +266,22 @@ object InferPType {
-      case _ => throw new RuntimeException(s"unsupported node:\n${Pretty(node)}")
+      case ShuffleWith(keyFields, rowType, rowEType, keyEType, name, writer, readers) =>
+        val r = requiredness(node)
+        readers.pType.setRequired(r.required)
```
@catoverdrive ?

cseed commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -847,6 +868,123 @@ class Emit[C](
+        val shuffleType = x.shuffleType
+        val shufflePType = x.shufflePType
+
+        val (socket, in, out, log) = ShuffleClient.openConnection(cb, shuffleType)
```
I'm having trouble commenting in our previous threads, so I'll ask this again here: it seems like you've inlined (part of) the shuffle client constructor and the implementation of ShuffleClient.start and ShuffleClient.stop.  I'm wondering why this code isn't just:

```
  val shuffleClient = CodeShuffleClient.create(cb, getPType(shuffleType))
  shuffleClient.start(cb)
  ...
  ...withCleanup {
    shuffleClient.stop(cb)
  }
```

danking replied:
ok

danking commented:
I think I still don't understand Requiredness.scala. I feel like this code has now gone full circle from pre-Requiredness days when I just wrote what I meant in InferPType, to putting it all in Requiredness and referencing it in InferPType, and now back to Requiredness mostly a formality and then I'm just writing what I want in InferPType.

catoverdrive commented on hail/src/main/scala/is/hail/expr/ir/Requiredness.scala:
```diff
@@ -96,6 +96,14 @@ class Requiredness(val usesAndDefs: UsesAndDefs, ctx: ExecuteContext) {
+      if (refMap.contains(name)) {
+        val uses = refMap(name)
+        uses.foreach { u => defs.bind(u, Array(req)) }
+        dependents.getOrElseUpdate(d, mutable.Set[RefEquality[BaseIR]]()) ++= uses
```
In this case, since the type of the `name` refs you're binding don't depend on any of the requiredness computation in the `ShuffleWith` node (They're always `TShuffle(required)`), they don't need to be added as dependents of the base (defining) node here since we don't need to recompute subtrees containing these refs if the ShuffleWith node changes. I think in this case I'd probably just inline this block in the ShuffleWith case, since it becomes a non-generic binding function.

catoverdrive replied:
(It's not going to yield incorrect results if you leave it in, you may just spend time unnecessarily computing subtrees when nothing has changed in the underlying requirednesses)

danking replied:
OK, addressed this.

danking commented on hail/src/main/scala/is/hail/services/shuffler/Base64Code.scala:
```diff
@@ -0,0 +1,26 @@
+
+class Base64EncoderValue(
+  code: Value[Base64.Encoder]
+) extends Value[Base64.Encoder] {
```
OK, I understand. I'll align to current conventions.

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -241,6 +244,21 @@ case class IEmitCode(Lmissing: CodeLabel, Lpresent: CodeLabel, pc: PCode) {
 
+  def withCleanup(cb: EmitCodeBuilder)(emitCleanup: CodeBuilder => Unit): IEmitCode = {
+    val cleanupMethod = cb.mb.genMethod("cleanup", FastIndexedSeq(), UnitInfo)
+    cleanupMethod.emit(CodeBuilder.scopedVoid(cleanupMethod)(emitCleanup))
```
OK, I used your suggestion. Can you verify that my change to `IEmitCode.consume` makes sense? I needed a way to access the memoized field.

danking commented on hail/src/main/scala/is/hail/types/physical/PCanonicalShuffle.scala:
```diff
@@ -0,0 +1,76 @@
+      pt.representation.allocate(region, Wire.ID_SIZE))
+    cb.append(pt.representation.store(off, bytes))
+    new PCanonicalShuffleSettable(pt, off)
+  }
```
can you look this over? I had to use newField. I am not sure if I'm still supposed to do that.

tpoterba replied:
you had to use newField to prevent bytecode verification issues?

danking replied:
yes.

danking commented:
bump-a-bump-bump @cseed

danking commented:
![image of sea creature gently bumping into a glass wall](https://media.giphy.com/media/toe6vBrtWgKvKW7Ehz/giphy.gif)

danking commented:
@tpoterba Cotton neglected to do a final pass before he went on vacation. I really don't want this to lie fallow for two weeks given how much rebase hell I've gone through already. Can you take a quick scan to ensure you don't see anything egregiously wrong? I am confident I addressed cotton's concerns.

My only concern is how I implemented `PCanonicalShuffle.fromArrayBytes` https://github.com/hail-is/hail/pull/8726/files#diff-7a0a59bff46d1da4904d669167c87f9cR34

tpoterba commented on hail/src/main/scala/is/hail/lir/X.scala:
```diff
@@ -69,11 +69,10 @@ class Classx[C](val name: String, val superName: String) {
           if (l.method == null)
             l.method = m
           else {
-            /*
```
change back

tpoterba commented on hail/src/main/scala/is/hail/lir/X.scala:
```diff
@@ -329,7 +328,7 @@ class MethodLit(
 class Local(var method: Method, val name: String, val ti: TypeInfo[_]) {
   override def toString: String = f"t${ System.identityHashCode(this) }%08x/$name ${ ti.desc }"
-  // val stack = Thread.currentThread().getStackTrace
+  val stack = Thread.currentThread().getStackTrace
```
change b ack

tpoterba commented on hail/src/main/scala/is/hail/types/physical/PCanonicalShuffle.scala:
```diff
@@ -0,0 +1,76 @@
+  }
+}
+
+class PCanonicalShuffleSettable(val pt: PCanonicalShuffle, shuffle: Settable[Long]) extends PShuffleValue with PSettable {
```
can this class wrap a PCanonicalBinarySettable instead of a Settable[Long]? It makes me uncomfortable that we could break this if we changed PCanonicalBinary.

danking replied:
done

tpoterba commented on hail/src/main/scala/is/hail/types/physical/PCanonicalShuffle.scala:
```diff
@@ -0,0 +1,76 @@
+  def store(pc: PCode): Code[Unit] = shuffle.store(pc.asInstanceOf[PCanonicalShuffleCode].shuffle)
+}
+
+class PCanonicalShuffleCode(val pt: PCanonicalShuffle, val shuffle: Code[Long]) extends PShuffleCode {
```
same.

danking replied:
done

tpoterba commented on hail/src/main/scala/is/hail/types/physical/PShuffle.scala:
```diff
@@ -0,0 +1,33 @@
+
+  def loadBytes(): Code[Array[Byte]]
+
+  def loadByte(i: Code[Int]): Code[Byte]
```
seems unused?

tpoterba commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -885,6 +892,77 @@ class Emit[C](
+        val isMissing = cb.newLocal[Boolean]("shuffleWithIsMissing")
+
+        val shuffleReaders = emitI(readersIR, env = shuffleEnv).consume(cb,
+          { cb.append(isMissing := const(true)) },
```
appears in a few places.

danking replied:
EmitCodeBuilder does not have store

danking commented on hail/src/main/scala/is/hail/expr/ir/Emit.scala:
```diff
@@ -885,6 +892,77 @@ class Emit[C](
+
+        val shuffleEnv = env.bind(name -> mb.newPresentEmitSettable(uuid.pt, uuid))
+
+        val successfulShuffleIds: PValue = emitI(writerIR, env = shuffleEnv).consume(cb,
```
Yes, see comment below. Eventually this will be used to ensure that the shuffler only preserves the one attempt per task. That functionality is not yet implemented.
